---
title: "TissueEnrich: A tool to calculate tissue-specific gene enrichment"
author: |
    | Ashish Jain, Geetu Tuteja
    | Bioinformatics and Computational Biology
    | Genetics, Development, and Cell Biology
    | Iowa State University, Ames, Iowa
    | 
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    #number_sections: true  ## if you want number sections at each table header
    #theme: united  # many options for theme, this one is my favorite.
bibliography: library.bib
fig_width: 8 
fig_height: 5 
vignette: >
  %\VignetteIndexEntry{TissueEnrich}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width=8, fig.height=5
)
```

# TissueEnrich
The `TissueEnrich` package is used to calculate enrichment of tissue-specific gene groups, given a set of input genes. For example, the user can input the most highly expressed genes from RNA-Seq data, or gene co-expression modules, to determine which tissue-specific genes are enriched in those datasets. Tissue-specific gene groups were generated by processing RNA-Seq data from the Human Protein Atlas (HPA) [@HPA2015], GTEx [@GTEx2015], and mouse ENCODE projects [@ENCODE2012], using an algorithm similar to the one used by the HPA project [@HPA2015]. The hypergeometric test is being used to determine if the tissue-specific genes are enriched among the input genes. Along with tissue-specific gene enrichment, `TissueEnrich` package can also be used to calculate various tissue-specific gene group in the user input expression datasets, which can further be used to carry out tissue-specific gene enrichment. `TissueEnrich` has the following three functions.

* `teEnrichment`: Given the input gene set, calculates the tissue-specific gene enrichment in human and mouse input genes.
* `teGeneRetrieval`: Given the expression data across tissues, calculates the tissue-specific genes by using the algorithm from the HPA project.
* `teEnrichmentCustom`: Given the input gene set, calculates the tissue-specific gene enrichment using tissue-specific genes, identified from the `teGeneRetrieval` function.

# How to get help for TissueEnrich
Please post all the questions or queries related to TissueEnrich package on the **Bioconductor support website**. This will help us to build a information repository which can be used by other users.

[https://support.bioconductor.org](https://support.bioconductor.org)

Please **do not** email your questions directly to the package authors.

# `teEnrichment`: Tissue-specific gene enrichment in human and mouse

The `teEnrichment` function carries out the tissue-specific gene enrichment in the input gene set. It uses the tissue-specific gene groups identified by processing RNA-Seq datasets in human and mouse. The user can specify the organism by using `organism` argument (1 for "Homo Sapiens" (default) and 2 for "Mus Musculus"). More details about the RNA-Seq datasets and tissue-specific groups are discussed in the next sections.

## RNA-Seq datasets
`TissueEnrich` consists of tissue-specific genes identified from widely popular and studied RNA-Seq datasets, having expression data across a variety of tissues. In order to make the tissue-specific genes more robust, tissues only with expression data from more than or equal to 2 biological replicates are selected. The datasets used in the tool are:

* **Human Protein Atlas:** RNA-Seq data across 35 human tissues [@HPA2015].
* **GTEx Dataset:** RNA-Seq data across 29 human tissues [@GTEx2015].
* **Mouse ENCODE Dataset:** RNA-Seq data across 17 mouse tissues [@ENCODE2012].

In `teEnrichment`, the user can specify the RNA-Seq dataset (`rnaSeqDataset`) to be used, for the tissue-specific gene enrichment analysis.

* 1 for "Human Protein Atlas" (default)
* 2 for "GTEx"
* 3 for "Mouse ENCODE"

## Tissue-specific Groups
Tissue specificity in the input gene set is determined based on the enrichment of tissue-specific genes. The tissue-specific genes are calculated based on the algorithm used in HPA project [@HPA2015]. The tissue-specific groups are defined as:

* **Tissue Enriched:** Genes whose mRNA expression is greater than 1 (TPM/FPKM) and at least five-fold higher mRNA levels in a particular tissue as compared to all other tissues.
* **Group Enriched:** Genes whose mRNA expression is greater than 1 (TPM/FPKM) and at least five-fold higher mRNA levels in a group of 2-7 tissues and does not come in any of the above 2 groups.
* **Tissue Enhanced:** Genes whose mRNA expression is greater than 1 (TPM/FPKM) and at least five-fold higher mRNA levels in a particular tissue as compared to average levels in all other tissues and does not come in any of the above 2 groups.

In `teEnrichment`, the user can specify the tissue-specific gene group (`tissueSpecificGeneType`) for the tissue-specific gene enrichment analysis.

* 1 for "All" (default)
* 2 for "Tissue-Enriched"
* 3 for "Tissue-Enhanced"
* 4 for "Group-Enriched"

## Hypergeometric test
The hypergeometric test is used for carrying out the tissue-specific gene enrichment. The p-value is calculated as:

$$ P(X \gt k) = \sum\limits_{i=k+1}^n \frac{{{K}\choose{i}} {{N-K}\choose{n-i}}}{{{N}\choose{n}}} $$
Where, N is the total number of genes, K is the total number of tissue-specific genes for a tissue, n is the number of genes in the input gene set, k is the number of tissue-specific genes in the input gene set. The p-values can be corrected for the multiple hypothesis testing, using Bonferroni Hochberg correction, by setting `multiHypoCorrection = TRUE`.

## Example: Tissue-specific gene enrichment
For our example, we used trophectoderm (TE) specific genes identified from single cell RNA-Seq analyses, performed on human blastocysts on days 5, 6, and 7 of preimplantation development. By using PCA, the single cells are assigned to either the inner cell mass (epiblast plus emerging extraembryonic endoderm) or the TE. After that, by using the differential expression analysis, a list of 100 TE-specific genes was generated [@Petropoulos2016]. We used those genes as the input gene set to check the tissue-specific enrichment in HPA. 

**Note:** The user can enter input gene set using either Ensembl Id or Gene Symbol (`geneFormat`).

```{r eval=TRUE}
library(TissueEnrich)
library(dplyr)
genes<-system.file("extdata", "inputGenes.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichment(inputGenes = inputGenes,geneFormat=2)
```

The `output` is a list object containing the enrichment results. These results are explained in the next section.

### Exploring tissue-specific gene enrichment results

#### Tissue-specific gene enrichment histogram using `ggplot2`
The first object in the `output` list is a dataframe containing the $-Log_{10}(P-Value)$ along with the number of tissue-specific genes for each tissue in the input gene list. This dataframe object can be used to visualize tissue-specific gene enrichment in the form of a histogram.

```{r eval=TRUE}
library(ggplot2)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

In the plot above, the x-axis represents the tissues and the y-axis represents the $-Log_{10}(P-Value)$ value. As expected, the 100 TE-specific genes shows enrichment of placenta specific genes.

#### Heatmap to show expression profiles of tissue-specific genes using `ggplot2`
The second object in the `output` list is a list containing the tissue-specific genes (with expression and tissue-specificity information) found in each tissue. The first item in this list has expression values of the tissue-specific genes identified in the input genes across all the tissues. The expression values can be visualized in the form of a heatmap. For example, the code below shows the placenta tissue-specific genes expression across all the tissues.

```{r eval=TRUE}
library(tidyr)
exp<-output[[2]][["Placenta"]][[1]]
exp$Gene<-row.names(exp)
exp<-exp %>% gather(Tissue=1:(ncol(exp)-1))

ggplot(exp, aes(key, Gene)) + geom_tile(aes(fill = value),
     colour = "white") + scale_fill_gradient(low = "white",
     high = "steelblue")+
     labs(x='', y = '')+
      theme_bw()+
      guides(fill = guide_legend(title = "Log2(TPM)"))+
      #theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())

```

#### Retrieval of input tissue-specific genes group
As mentioned above, the second object in the `output` list also consists of information about the tissue-specificity information of the tissue-specific genes identified in the input genes, along with the expression data. The code below retrieves the tissue-specific genes along with their tissue-specific group in the placenta.

```{r eval=TRUE}
groupInf<-output[[2]][["Placenta"]][[2]]
print(head(groupInf))
```


#### Retrieval of tissue-specific genes not found in the tool
The third object in the `output` list is a character vector consisting of a list of input genes that are not in our dataset.
```{r eval=TRUE}
print(output[[3]])
```

## Orthologous gene enrichment
This `teEnrichment` function also has the functionality of calculating mouse tissue-specific gene enrichment in the input human gene list or vice versa. For using this, the user just has to use mouse dataset for the human input genes or vice versa. The function will automatically carry out orthologous tissue-specific gene enrichment using one-to-one protein coding orthologous genes between human and mouse, downloaded from Ensembl V91 database [@Aken2016].

### Example: Tissue-specific gene enrichment of mouse tissues in input human genes
In this example, the list of 100 TE-specific genes from the previous example is used to carry out tissue-specific gene enrichment using mouse ENCODE data.

```{r eval=TRUE}
genes<-system.file("extdata", "inputGenes.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichment(inputGenes = inputGenes,rnaSeqDataset = 3, organism = 1, geneFormat=2)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

As expected, the plot shows enrichment of placenta specific genes in the input list of 100 human TE-specific genes.

# `teGeneRetrieval`: Identification of tissue-specific genes

The `teGeneRetrieval` function divides the genes into tissue-specific groups, based on the algorithm used in HPA project [@HPA2015]. It takes gene expression matrix as the input (rows as genes and columns as tissue) and classifies them into different gene groups. The users also have the option to change the thresholds (other than in HPA) to vary the degree of tissue specificity of genes. More information about the gene groups and HPA thresholds is described below.

## Gene groups
The genes are divided into six groups based on their gene expression across the tissues. These groups are:

* **Not Detected:** Genes whose mRNA expression is less than 1 (TPM/FPKM) across all the tissues.
* **Tissue Enriched:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a particular tissue as compared to all other tissues.
* **Group Enriched:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a group of 2-7 tissues and doesn't come in any of the above 2 groups.
* **Tissue Enhanced:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a particular tissue as compared to average levels in all tissues and doesn't come in any of the above 3 groups.
* **Expressed in all:** Genes whose mRNA expression is greater than or equal to 1 (TPM/FPKM) across all the tissues and doesn't come in any of the above 4 groups.
* **Mixed:** Genes that donâ€™t occur in any of the above 5 groups.

Out of these groups, the **Tissue Enriched**, **Group Enriched**, and **Tissue Enhanced** groups are classified as tissue-specific.

## Example: Tissue-specific gene retrieval
In the example below, we supplied a subset of mouse ENCODE data, consisting expression data of 36 genes across 17 tissues.

```{r eval=TRUE}
library(TissueEnrich)
data<-system.file("extdata", "test.expressiondata.txt", package = "TissueEnrich")
expressionData<-read.table(data,header=TRUE,row.names=1,sep='\t')
TSgenes<-teGeneRetrieval(expressionData)
head(TSgenes)
```

As seen above, the output is a data frame with columns having information about Gene name, Tissue name, and Tissue-Specific group.

# `teEnrichmentCustom`: Tissue-specific gene enrichment in custom expression datasets

The `teEnrichmentCustom` function calculates the tissue-specific gene enrichment, using the tissue-specific gene groups information from custom expression datasets, calculated from `teGeneRetrieval` function.

## Example: Tissue-specific gene enrichment in custom dataset
In the example below, we used 10 genes, randomly selected from the 36 genes used in the previous example. The tissue-specific genes identified from the custom gene expression can be used to calculate tissue-specific gene enrichment in the input gene set.

```{r eval=FALSE}
genes<-system.file("extdata", "inputGenesEnsembl.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichmentCustom(inputGenes,TSgenes)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

# References

[1]: https://www.proteinatlas.org/humanproteome/tissue+specific
[2]: https://www.gtexportal.org/home/documentationPage
[3]: http://chromosome.sdsc.edu/mouse/download.html
[4]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4868821/
