---
title: "TissueEnrich: A tool to calculate tissue-specific gene enrichment"
author: |
    | Ashish Jain, Geetu Tuteja
    | Bioinformatics and Computational Biology
    | Genetics, Development, and Cell Biology
    | Iowa State University, Ames, Iowa
    | 
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: >
  Cell clusters from single-cell RNA-Seq data and differentiated embryonic 
  stem cells are often distinguished based on the expression of a few marker 
  genes, even though such markers can have a similar expression in more than 
  one cell type. Therefore, there is a need for a method to identify these 
  cells more robustly. Taking this into account, we developed TissueEnrich, 
  a tool to identify tissue specificity of cells, based on the enrichment of 
  tissue-specific genes in these cells. This tool consists of a user-friendly 
  and interactive web application as well as an R package to calculate 
  tissue-specific gene enrichment. 
  TissueEnrich package version: `r packageVersion("TissueEnrich")`
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    #number_sections: true  ## if you want number sections at each table header
    #theme: united  # many options for theme, this one is my favorite.
bibliography: library.bib
fig_width: 8 
fig_height: 5 
vignette: >
  %\VignetteIndexEntry{TissueEnrich}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width=8, fig.height=5
)
```

# `TissueEnrich`
The `TissueEnrich` package is used to calculate enrichment of tissue-specific gene groups, given a input genes. For example, the user can input the most highly expressed genes from RNA-Seq data, or gene co-expression modules, to determine which tissue-specific genes are enriched in those datasets. Tissue-specific gene groups were generated by processing RNA-Seq data from the Human Protein Atlas (HPA) [@HPA2015], GTEx [@GTEx2015], and mouse ENCODE projects [@ENCODE2012], using an algorithm similar to the one used by the HPA project [@HPA2015]. The hypergeometric test is being used to determine if the tissue-specific genes are enriched in the input genes or not. Along with tissue-specific gene enrichment, `TissueEnrich` package can also be used to calculate various tissue-specific gene group in the input expression datasets, which can further be used to carry out tissue-specific gene enrichment. `TissueEnrich` has the following three functions.

* `teEnrichment`: Given the input gene set, calculates the tissue-specific gene enrichment in mouse and humans.
* `teGeneRetrieval`: Given the expression data across tissues, calculates the tissue-specific genes by using the algorithm from the HPA project.
* `teEnrichmentCustom`: Given the input gene set, calculates the tissue-specific gene enrichment using tissue-specific genes calculated from custom datasets.

# How to get help for TissueEnrich
Please post all the questions or queries related to TissueEnrich package on the **Bioconductor support website**. This will help other users to browse through the solutions if they face similar problems.

[https://support.bioconductor.org](https://support.bioconductor.org)

Please **do not** email your questions directly to the package authors.

# Using `teEnrichment`

The `teEnrichment` function carries out the tissue-specific gene enrichment in the input gene set. It uses the tissue-specific gene groups identified by processing RNA-Seq studies in human and mouse to calculate enrichment. More information about the various tissue-specific groups and RNA-Seq studies are specified in the following sections.

## Tissue-specific Groups
Tissue specificity in the input gene set is determined based on the enrichment of tissue-specific genes in the above datasets. The tissue-specific genes in the datasets are calculated based on the thresholds specified in Human Protein Atlas [@HPA2015]. The tissue-specific groups are defined as:

* **Tissue Enriched:** Genes whose mRNA expression is greater than 1 (TPM/FPKM) and at least five-fold higher mRNA levels in a particular tissue as compared to all other tissues.
* **Group Enriched:** Genes whose mRNA expression is greater than 1 (TPM/FPKM) and at least five-fold higher mRNA levels in a group of 2-7 tissues and does not come in any of the above 2 groups.
* **Tissue Enhanced:** Genes whose mRNA expression is greater than 1 (TPM/FPKM) and at least five-fold higher mRNA levels in a particular tissue as compared to average levels in all tissues and does not come in any of the above 2 groups.

## RNA-Seq datasets
The tissue-specific gene enrichment can be 

* **Human Protein Atlas:** RNA-Seq data across 35 human tissues.
* **GTEx Dataset:** RNA-Seq data across 29 human tissues.
* **Mouse ENCODE Dataset:** RNA-Seq data across 17 mouse tissues.

## Enrichment using hypergeometric test
The user can specify (`tissueSpecificGeneType`) which tissue-specific gene group to be used for carrying out enrichment. The user can also enter input gene set in two different formats (Ensembl Id or Gene Symbol). The tissue-specific gene enrichment is calculated using hypergeometric distribution.

$$ P(X \gt k) = \sum\limits_{i=k+1}^n \frac{{{K}\choose{i}} {{N-K}\choose{n-i}}}{{{N}\choose{n}}} $$
Here, N is the total number of genes, K is the total number of tissue-specific genes for a tissue, n is the number of genes in the input gene set, k is the number of tissue-specific genes in the input gene set. The p-values were also corrected for the multiple hypotheses, using Bonferroni Hochberg correction. There is also a flag to carry out multiple hypothesis correction using Bonferroni Hochberg method. 

## Example: Tissue-specific gene enrichment 
In the example below, we used a gene list containing the top 100 polar genes found in a human single cell dataset [4].

```{r eval=TRUE}
library(TissueEnrich)
library(dplyr)
genes<-system.file("extdata", "inputGenes.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichment(inputGenes = inputGenes,geneFormat=2)
```

The output from this function is a list object consist of three objects. The first object is a dataframe object containing the -Log10(P-Value) for each tissue in the dataset. This dataframe object can be used to plot a histogram as shown below.

### Exploring tissue-specific gene enrichment results

#### Tissue-specific gene enrichment histogram using `ggplot2`
```{r eval=TRUE}
library(ggplot2)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

#### Retrieval of tissue-specific genes found in the input gene set
The second object is a list containing the tissue-specific genes (with expression and tissue-specificity information) found in each tissue. For example, the code below retrieves the tissue-specific genes with their tissue-specificity in placenta.

```{r eval=TRUE}
print("Placenta specific genes found in input genes with group specificity:")
groupInf<-output[[2]][["Placenta"]][[2]]
print(head(groupInf))
```

#### Heatmaps to show expression profiles of tissue-specific genes using `ggplot2`
It can be used to plot the heatmap showing the expression values of the tissue-specific genes across all the tissues. For example, the code below plots the placenta tissue-specific genes expression across all the tissues.
```{r eval=TRUE}
library(tidyr)
exp<-output[[2]][["Placenta"]][[1]]
exp$Gene<-row.names(exp)
exp<-exp %>% gather(Tissue=1:(ncol(exp)-1))

ggplot(exp, aes(key, Gene)) + geom_tile(aes(fill = value),
     colour = "white") + scale_fill_gradient(low = "white",
     high = "steelblue")+
     labs(x='', y = '')+
      theme_bw()+
      guides(fill = guide_legend(title = "Log2(TPM)"))+
      #theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())

```

#### Retrieval of tissue-specific genes not found in the tool
The third object is a character vector consisting of a list of input genes that are not in our dataset.
```{r eval=TRUE}
print("Input genes not found in our dataset are:")
print(output[[3]])
```

## Example: Homologous Use 
This `teEnrichment` function also has the functionality of using a mouse dataset to calculate tissue-specific gene enrichment in a input human gene list or vice versa. The `isHomolog` flag need to be set `TRUE` to use this functionality. This argument uses one-to-one protein coding homologs to map human genes to mouse genes and vice versa using Ensembl V91 database. For example, the above used human input genes can be used with Mouse ENCODE dataset.

```{r eval=TRUE}
genes<-system.file("extdata", "inputGenes.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichment(inputGenes = inputGenes,rnaSeqDataset = 3, organism = 1, geneFormat=2,isHomolog = TRUE)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

# Using `teGeneRetrieval`

This `teGeneRetrieval` function calculates the tissue-specific genes based on the algorithm from the Human Protein Atlas project [1]. 

## Tissue specific gene groups
The genes are divided into six groups based on their gene expression across the tissues. These groups are:

* **Not Detected:** Genes whose mRNA expression is less than 1 (TPM/FPKM) across all the tissues.
* **Tissue Enriched:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a particular tissue as compared to all other tissues.
* **Group Enriched:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a group of 2-7 tissues and doesn't come in any of the above 2 groups.
* **Tissue Enhanced:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a particular tissue as compared to average levels in all tissues and doesn't come in any of the above 3 groups.
* **Expressed in all:** Genes whose mRNA expression is greater than or equal to 1 (TPM/FPKM) across all the tissues and doesn't come in any of the above 4 groups.
* **Mixed:** Genes that don’t occur in any of the above 5 groups.

Out of these, **Tissue Enriched**, **Group Enriched**, and **Tissue Enhanced** groups are classified as tissue-enriched.

## Example: Tissue-specific gene retrieval
The function `teGeneRetrieval` takes gene expression matrix as the input (rows as genes and columns as tissue) and calculates the tissue-specific genes. Although, the above mentioned criterias are specified in Human Protein Atlas [1], users can also change the different thresholds to calculate tissue-specific genes in their dataset. The output consists of a data frame with the following columns: Gene Name, Tissue Name, and Tissue-Specific Group.


```{r eval=TRUE}
library(TissueEnrich)
data<-system.file("extdata", "test.expressiondata.txt", package = "TissueEnrich")
expressionData<-read.table(data,header=TRUE,row.names=1,sep='\t')
TSgenes<-teGeneRetrieval(expressionData)
head(TSgenes)
```

# Using `teEnrichmentCustom`

The `teEnrichmentCustom` function is used to calculate tissue-specific gene enrichment using the processed data calculated from `teGeneRetrieval` function.

## Example: Tissue-specific gene enrichment in custom dataset
```{r eval=FALSE}
data<-system.file("extdata", "test.expressiondata.txt", package = "TissueEnrich")
expressionData<-read.table(data,header=TRUE,row.names=1,sep='\t')
TSgenes<-teGeneRetrieval(expressionData)
genes<-system.file("extdata", "inputGenesEnsembl.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichmentCustom(inputGenes,TSgenes)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

# References

[1]: https://www.proteinatlas.org/humanproteome/tissue+specific
[2]: https://www.gtexportal.org/home/documentationPage
[3]: http://chromosome.sdsc.edu/mouse/download.html
[4]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4868821/
