---
title: "TissueEnrich"
output: rmarkdown::html_vignette
fig_width: 6 
fig_height: 4 
vignette: >
  %\VignetteIndexEntry{TissueEnrich}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width=7, fig.height=5
)
```

`TissueEnrich` package is used to calculate tissue-specific gene enrichment in the input gene groups. The user can input the top expressed gene set from single-cell RNA-Seq data or co-expressed gene modules to identify the tissue-specificity in those datasets. It has the tissue-specific genes processed from large-scale RNA-Seq data of various tissues studies from Human protein atlas, GTEx, and mouse ENCODE data. The tissue-specific genes are calculated based on the algorithm from human protein atlas project [1]. The hypergeometric test is being used to calculate the tissue-specific gene enrichment in the input genes. `TissueEnrich` consists of three functions to calculate the tissue-specific genes and tissue-specific gene enrichment.

* `teEnrichment`: Calculation of tissue-specific gene enrichment in mouse and humans using hypergeometric test.
* `teGeneRetrieval`: Calculation of tissue-specific genes by using the algorithm from the Human Protein Atlas project.
* `teEnrichmentCustom`: Calculation of tissue-specific gene enrichment using custom datasets.

# Using `teEnrichment`

The `teEnrichment` function is used to calculate tissue-specific gene enrichment using the processed data from widely used RNA-Seq studies in human and mouse. It has following RNA-Seq datasets:

* **Human Protein Atlas [1]:** RNA-Seq data across 35 human tissues.
* **GTEx Dataset [2]:** RNA-Seq data across 29 human tissues.
* **Mouse ENCODE Dataset [3]:** RNA-Seq data across 17 mouse tissues.

Tissue-specificity in a user’s gene set will be determined based on the enrichment of tissue-specific genes in the above datasets. The tissue-specific genes in the datasets are calculated based on the thresholds specified in Human Protein Atlas. The tissue-specific gene enrichement is carried out using hypergeometric distribution.

$$ P(X \gt k) = \sum\limits_{i=k+1}^n \frac{{{K}\choose{i}} {{N-K}\choose{n-i}}}{{{N}\choose{n}}} $$
Here, N is the total number of genes, K is the total number of tissue-specific genes for a tissue, n is the number of genes in the input gene set, k is the number of tissue-specific genes in the input gene set. The p-values were also corrected for the multiple hypotheses, using Bonferroni Hochberg correction.

There is also a flag to carry out multiple hypothesis correction using Bonferroni Hochberg method. In the example below, we used a gene list containing the top 100 polar genes found in a human single cell dataset [4].

```{r eval=TRUE}
library(TissueEnrich)
library(dplyr)
genes<-system.file("extdata", "inputGenes.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichment(inputGenes = inputGenes,geneFormat=2)
```

The output from this function is a list object consist of three objects. The first object is a dataframe object containing the -Log10(P-Value) for each tissue in the dataset. This dataframe object can be used to plot a histogram as shown below.

```{r eval=TRUE}
library(ggplot2)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

The second object is a list containing the tissue-specific genes (with expression and tissue-specificity information) found in each tissue. For example, the code below retrieves the tissue-specific genes with their tissue-specificity in placenta.

```{r eval=TRUE}
print("Placenta specific genes found in input genes with group specificity:")
groupInf<-output[[2]][["Placenta"]][[2]]
print(head(groupInf))
```

It can be used to plot the heatmap showing the expression values of the tissue-specific genes across all the tissues. For example, the code below plots the placenta tissue-specific genes expression across all the tissues.
```{r eval=TRUE}
library(tidyr)
exp<-output[[2]][["Placenta"]][[1]]
exp$Gene<-row.names(exp)
exp<-exp %>% gather(Tissue=1:(ncol(exp)-1))

ggplot(exp, aes(key, Gene)) + geom_tile(aes(fill = value),
     colour = "white") + scale_fill_gradient(low = "white",
     high = "steelblue")+
     labs(x='', y = '')+
      theme_bw()+
      guides(fill = guide_legend(title = "Log2(TPM)"))+
      #theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())

```

The third object is a character vector consisting of a list of input genes that are not in our dataset.
```{r eval=TRUE}
print("Input genes not found in our dataset are:")
print(output[[3]])
```

This `teEnrichment` function also has the functionality of using a mouse dataset to calculate tissue-specific gene enrichment in a input human gene list or vice versa. The `isHomolog` flag need to be set `TRUE` to use this functionality. This argument uses one-to-one protein coding homologs to map human genes to mouse genes and vice versa using Ensembl V91 database. For example, the above used human input genes can be used with Mouse ENCODE dataset.

```{r eval=TRUE}
genes<-system.file("extdata", "inputGenes.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichment(inputGenes = inputGenes,rnaSeqDataset = 3, organism = 1, geneFormat=2,isHomolog = TRUE)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```

# Using `teGeneRetrieval`

This `teGeneRetrieval` function calculates the tissue-specific genes based on the algorithm from the Human Protein Atlas project [1]. The genes are divided into six groups based on their gene expression across the tissues. These groups are:

* **Not Detected:** Genes whose mRNA expression is less than 1 (TPM/FPKM) across all the tissues.
* **Tissue Enriched:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a particular tissue as compared to all other tissues.
* **Group Enriched:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a group of 2-7 tissues and doesn't come in any of the above 2 groups.
* **Tissue Enhanced:** Genes whose mRNA expression is at least five-fold higher mRNA levels in a particular tissue as compared to average levels in all tissues and doesn't come in any of the above 3 groups.
* **Expressed in all:** Genes whose mRNA expression is greater than or equal to 1 (TPM/FPKM) across all the tissues and doesn't come in any of the above 4 groups.
* **Mixed:** Genes that don’t occur in any of the above 5 groups.

Out of these, **Tissue Enriched**, **Group Enriched**, and **Tissue Enhanced** groups are classified as tissue-enriched.

The function `teGeneRetrieval` takes gene expression matrix as the input (rows as genes and columns as tissue) and calculates the tissue-specific genes. Although, the above mentioned criterias are specified in Human Protein Atlas [1], users can also change the different thresholds to calculate tissue-specific genes in their dataset. The output consists of a data frame with the following columns: Gene Name, Tissue Name, and Tissue-Specific Group.


```{r eval=TRUE}
library(TissueEnrich)
data<-system.file("extdata", "test.expressiondata.txt", package = "TissueEnrich")
expressionData<-read.table(data,header=TRUE,row.names=1,sep='\t')
TSgenes<-teGeneRetrieval(expressionData)
head(TSgenes)
```

# Using `teEnrichmentCustom`

The `teEnrichmentCustom` function is used to calculate tissue-specific gene enrichment using the processed data calculated from `teGeneRetrieval` function.

```{r eval=FALSE}
data<-system.file("extdata", "test.expressiondata.txt", package = "TissueEnrich")
expressionData<-read.table(data,header=TRUE,row.names=1,sep='\t')
TSgenes<-teGeneRetrieval(expressionData)
genes<-system.file("extdata", "inputGenesEnsembl.txt", package = "TissueEnrich")
inputGenes<-scan(genes,character())
output<-teEnrichmentCustom(inputGenes,TSgenes)
ggplot(output[[1]],aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity')+
      labs(x='', y = '-LOG10(P-Value)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
```


[1]: https://www.proteinatlas.org/humanproteome/tissue+specific
[2]: https://www.gtexportal.org/home/documentationPage
[3]: http://chromosome.sdsc.edu/mouse/download.html
[4]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4868821/
